### PR TITLE
[Feature] user apps

### PR DESCRIPTION
**Pull request by Misha-133**
### Description
This PR adds initial support for the new User Apps feature

### Changes
- new command properties
- - new Interaction Framework attributes
- new `IDiscordInteraction` properties
- new `IMessage` properties
- [x] Docs page covering the feature

### CODE REVIEW

File: src/Discord.Net.Core/Entities/Applications/ApplicationIntegrationType.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+namespace Discord;
+
+public enum ApplicationIntegrationType
```
Comment (quinchs) [inline src/Discord.Net.Core/Entities/Applications/ApplicationIntegrationType.cs:6]: nit: docs

File: src/Discord.Net.Core/Entities/Interactions/ApplicationCommands/IApplicationCommand.cs [hunk 2]
```diff
@@ -83,6 +84,16 @@ public interface IApplicationCommand : ISnowflakeEntity, IDeletable
         /// </remarks>
         string DescriptionLocalized { get; }
 
+        /// <summary>
+        ///     Gets context types the command can be used in. <see langword="null" /> if not specified.
```
Comment (quinchs) [inline src/Discord.Net.Core/Entities/Interactions/ApplicationCommands/IApplicationCommand.cs]:
```suggestion
        ///     Gets context types the command can be used in; <see langword="null" /> if not specified.
```

File: src/Discord.Net.Core/Entities/Interactions/ApplicationCommands/IApplicationCommand.cs [hunk 3]
```diff
@@ -83,6 +84,16 @@ public interface IApplicationCommand : ISnowflakeEntity, IDeletable
...
+        /// </summary>
+        IReadOnlyCollection<InteractionContextType> ContextTypes { get; }
+
+        /// <summary>
+        ///     Gets the install method for the command. <see langword="null" /> if not specified.
```
Comment (quinchs) [inline src/Discord.Net.Core/Entities/Interactions/ApplicationCommands/IApplicationCommand.cs]:
```suggestion
        ///     Gets the install method for the command; <see langword="null" /> if not specified.
```

File: src/Discord.Net.Core/Entities/Interactions/ApplicationCommands/InteractionContextType.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+namespace Discord;
+
+public enum InteractionContextType
```
Comment (quinchs) [inline src/Discord.Net.Core/Entities/Interactions/ApplicationCommands/InteractionContextType.cs:6]: nit: docs

File: src/Discord.Net.Core/Entities/Interactions/ContextMenus/MessageCommandBuilder.cs [hunk 5]
```diff
@@ -56,6 +57,16 @@ public string Name
         /// </summary>
         public GuildPermission? DefaultMemberPermissions { get; set; }
 
+        /// <summary>
+        ///     Gets the install method for this command.
```
Comment (quinchs) [inline src/Discord.Net.Core/Entities/Interactions/ContextMenus/MessageCommandBuilder.cs]:
```suggestion
        ///     Gets the install method for this command; <see langword="null" /> if not specified.
```

File: src/Discord.Net.Core/Entities/Interactions/ContextMenus/MessageCommandBuilder.cs [hunk 6]
```diff
@@ -56,6 +57,16 @@ public string Name
...
+        /// </summary>
+        public HashSet<ApplicationIntegrationType> IntegrationTypes { get; set; } = null;
+
+        /// <summary>
+        ///     Gets the context types this command can be executed in.
```
Comment (quinchs) [inline src/Discord.Net.Core/Entities/Interactions/ContextMenus/MessageCommandBuilder.cs]:
```suggestion
        ///     Gets the context types this command can be executed in; <see langword="null" /> if not specified.
```

File: src/Discord.Net.Core/Entities/Interactions/MessageInteractionData/IMessageInteractionData.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    /// </summary>
+    InteractionType Type { get; }
+
+    /// <summary>
+    ///     Gets the ID of the user who triggered the interaction.
+    /// </summary>
+    ulong UserId { get; }
+
+    /// <summary>
+    ///
```
Comment (quinchs) [inline src/Discord.Net.Core/Entities/Interactions/MessageInteractionData/IMessageInteractionData.cs]: missing docs

File: src/Discord.Net.Interactions/Attributes/CommandContextTypeAttribute.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Discord.Interactions;
+
+/// <summary>
+///     Specifies context types this command can be executed in.
+/// </summary>
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = false, Inherited = true)]
+public class CommandContextTypeAttribute : Attribute
+{
+    public IReadOnlyCollection<InteractionContextType> ContextTypes { get; }
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Attributes/CommandContextTypeAttribute.cs:17]: nit: docs

File: src/Discord.Net.Interactions/Attributes/CommandContextTypeAttribute.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+    public CommandContextTypeAttribute(params InteractionContextType[] contextTypes)
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Attributes/CommandContextTypeAttribute.cs:23]: docs

File: src/Discord.Net.Interactions/Attributes/IntegrationTypeAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Discord.Interactions;
+
+/// <summary>
+///     Specifies install method for the command.
+/// </summary>
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = false, Inherited = true)]
+public class IntegrationTypeAttribute : Attribute
+{
+    public IReadOnlyCollection<ApplicationIntegrationType> IntegrationTypes { get; }
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Attributes/IntegrationTypeAttribute.cs:17]: nit: docs

File: src/Discord.Net.Interactions/Attributes/IntegrationTypeAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+    public IntegrationTypeAttribute(params ApplicationIntegrationType[] integrationTypes)
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Attributes/IntegrationTypeAttribute.cs:23]: docs

File: src/Discord.Net.Interactions/Builders/Commands/ContextCommandBuilder.cs [hunk 12]
```diff
@@ -126,6 +144,28 @@ public ContextCommandBuilder WithDefaultMemberPermissions(GuildPermission permis
...
+        /// <param name="integrationTypes">Install types for this command.</param>
+        /// <returns>The builder instance.</returns>
+        public ContextCommandBuilder WithIntegrationTypes(params ApplicationIntegrationType[] integrationTypes)
+        {
+            IntegrationTypes = new HashSet<ApplicationIntegrationType>(integrationTypes);
+            return this;
+        }
+
+        /// <summary>
+        ///     Sets <see cref="ContextTypes"/>.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Builders/Commands/ContextCommandBuilder.cs]:
```suggestion
        ///     Sets the <see cref="ContextTypes"/> of this <see cref="ContextCommandBuilder"/>.
```

File: src/Discord.Net.Interactions/Builders/Commands/ContextCommandBuilder.cs [hunk 13]
```diff
@@ -126,6 +144,28 @@ public ContextCommandBuilder WithDefaultMemberPermissions(GuildPermission permis
             return this;
         }
 
+        /// <summary>
+        ///     Sets <see cref="IntegrationTypes"/>.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Builders/Commands/ContextCommandBuilder.cs]:
```suggestion
        ///     Sets the <see cref="IntegrationTypes"/> of this <see cref="ContextCommandBuilder"/>.
```

File: src/Discord.Net.Interactions/Builders/Commands/SlashCommandBuilder.cs [hunk 14]
```diff
@@ -35,7 +36,24 @@ public sealed class SlashCommandBuilder : CommandBuilder<SlashCommandInfo, Slash
         /// </summary>
         public GuildPermission? DefaultMemberPermissions { get; set; } = null;
 
-        internal SlashCommandBuilder(ModuleBuilder module) : base(module) { }
+        /// <summary>
+        ///     Gets the install method for this command.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Builders/Commands/SlashCommandBuilder.cs]:
```suggestion
        ///     Gets or sets the install method for this command.
```

File: src/Discord.Net.Interactions/Builders/Commands/SlashCommandBuilder.cs [hunk 15]
```diff
@@ -35,7 +36,24 @@ public sealed class SlashCommandBuilder : CommandBuilder<SlashCommandInfo, Slash
...
+        /// </summary>
+        public HashSet<ApplicationIntegrationType> IntegrationTypes { get; set; } = null;
+
+        /// <summary>
+        ///     Gets the context types this command can be executed in.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Builders/Commands/SlashCommandBuilder.cs]:
```suggestion
        ///     Gets or sets the context types this command can be executed in.
```

File: src/Discord.Net.Interactions/Builders/Commands/SlashCommandBuilder.cs [hunk 16]
```diff
@@ -126,6 +144,32 @@ public SlashCommandBuilder WithDefaultMemberPermissions(GuildPermission permissi
             return this;
         }
 
+        /// <summary>
+        ///     Sets <see cref="IntegrationTypes"/>.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Builders/Commands/SlashCommandBuilder.cs]:
```suggestion
        ///     Sets the <see cref="IntegrationTypes"/> on this <see cref="SlashCommandBuilder"/>.
```

File: src/Discord.Net.Interactions/Builders/Commands/SlashCommandBuilder.cs [hunk 17]
```diff
@@ -126,6 +144,32 @@ public SlashCommandBuilder WithDefaultMemberPermissions(GuildPermission permissi
...
+        public SlashCommandBuilder WithIntegrationTypes(params ApplicationIntegrationType[] integrationTypes)
+        {
+            IntegrationTypes = integrationTypes is not null
+                ? new HashSet<ApplicationIntegrationType>(integrationTypes)
+                : null;
+            return this;
+        }
+
+        /// <summary>
+        ///     Sets <see cref="ContextTypes"/>.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Builders/Commands/SlashCommandBuilder.cs]:
```suggestion
        ///     Sets the <see cref="ContextTypes"/> on this <see cref="SlashCommandBuilder"/>.
```

File: src/Discord.Net.Interactions/Builders/ModuleBuilder.cs [hunk 18]
```diff
@@ -114,6 +115,16 @@ public class ModuleBuilder
         /// </summary>
         public IReadOnlyList<ModalCommandBuilder> ModalCommands => _modalCommands;
 
+        /// <summary>
+        ///     Gets the install method for this command.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Builders/ModuleBuilder.cs]:
```suggestion
        ///     Gets or sets the install method for this command.
```

File: src/Discord.Net.Interactions/Builders/ModuleBuilder.cs [hunk 19]
```diff
@@ -114,6 +115,16 @@ public class ModuleBuilder
...
+        /// </summary>
+        public HashSet<ApplicationIntegrationType> IntegrationTypes { get; set; } = null;
+
+        /// <summary>
+        ///     Gets the context types this command can be executed in.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Builders/ModuleBuilder.cs]:
```suggestion
        ///     Gets or sets the context types this command can be executed in.
```

File: src/Discord.Net.Interactions/Builders/ModuleBuilder.cs [hunk 20]
```diff
@@ -423,6 +435,28 @@ public ModuleBuilder AddModule(Action<ModuleBuilder> configure)
             return this;
         }
 
+        /// <summary>
+        ///     Sets <see cref="IntegrationTypes"/>.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Builders/ModuleBuilder.cs]:
```suggestion
        ///     Sets the <see cref="IntegrationTypes"/> on this <see cref="ModuleBuilder"/>.
```

File: src/Discord.Net.Interactions/Builders/ModuleBuilder.cs [hunk 21]
```diff
@@ -423,6 +435,28 @@ public ModuleBuilder AddModule(Action<ModuleBuilder> configure)
...
+        /// <param name="integrationTypes">Install types for this command.</param>
+        /// <returns>The builder instance.</returns>
+        public ModuleBuilder WithIntegrationTypes(params ApplicationIntegrationType[] integrationTypes)
+        {
+            IntegrationTypes = new HashSet<ApplicationIntegrationType>(integrationTypes);
+            return this;
+        }
+
+        /// <summary>
+        ///     Sets <see cref="ContextTypes"/>.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Builders/ModuleBuilder.cs]:
```suggestion
        ///     Sets  the <see cref="ContextTypes"/> on this <see cref="ModuleBuilder"/>.
```

File: src/Discord.Net.Interactions/Info/IApplicationCommandInfo.cs [hunk 22]
```diff
@@ -37,5 +38,15 @@ public interface IApplicationCommandInfo
         ///     Gets the default permissions needed for executing this command.
         /// </summary>
         public GuildPermission? DefaultMemberPermissions { get; }
+
+        /// <summary>
+        ///     Gets context types this command can be executed in.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Info/IApplicationCommandInfo.cs]:
```suggestion
        ///     Gets the context types this command can be executed in.
```

File: src/Discord.Net.Interactions/Info/IApplicationCommandInfo.cs [hunk 23]
```diff
@@ -37,5 +38,15 @@ public interface IApplicationCommandInfo
...
+        /// </summary>
+        public IReadOnlyCollection<InteractionContextType> ContextTypes { get; }
+
+        /// <summary>
+        ///     Gets install methods for this command.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Info/IApplicationCommandInfo.cs]:
```suggestion
        ///     Gets the install methods for this command.
```

File: src/Discord.Net.Interactions/Info/ModuleInfo.cs [hunk 24]
```diff
@@ -117,6 +117,16 @@ public class ModuleInfo
         /// </summary>
         public bool DontAutoRegister { get; }
 
+        /// <summary>
+        ///     Gets context types commands in this module can be executed in.
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Info/ModuleInfo.cs]:
```suggestion
        ///     Gets the context types commands in this module can be executed in.
```

File: src/Discord.Net.Rest/Entities/RestApplication.cs [hunk 25]
```diff
@@ -6,183 +6,195 @@
...
+        if (response.Id != Id)
+            throw new InvalidOperationException("Unable to update this object from a different application token.");
+        Update(response);
     }
+
+    /// <summary>
+    ///     Gets the name of the application.
+    /// </summary>
+    /// <returns>
+    ///     Name of the application.
```
Comment (quinchs) [inline src/Discord.Net.Rest/Entities/RestApplication.cs]:
```suggestion
    ///     The name of the application.
```

**Review by quinchs — APPROVED** [review]

File: src/Discord.Net.Core/Entities/Applications/ApplicationIntegrationType.cs (hunk 1 above)
Comment (Misha-133) [inline src/Discord.Net.Core/Entities/Applications/ApplicationIntegrationType.cs]:
```suggestion

/// <summary>
///     Defines where an application can be installed.
/// </summary>
public enum ApplicationIntegrationType
```

File: src/Discord.Net.Core/Entities/Interactions/ApplicationCommands/InteractionContextType.cs (hunk 4 above)
Comment (Misha-133) [inline src/Discord.Net.Core/Entities/Interactions/ApplicationCommands/InteractionContextType.cs]:
```suggestion

/// <summary>
///     Represents a context in Discord where an interaction can be used.
/// </summary>
public enum InteractionContextType
```

File: src/Discord.Net.Core/Entities/Interactions/MessageInteractionData/IMessageInteractionData.cs (hunk 7 above)
Comment (Misha-133) [inline src/Discord.Net.Core/Entities/Interactions/MessageInteractionData/IMessageInteractionData.cs]:
```suggestion
    /// <summary>
    ///     Gets the Ids for installation contexts related to the interaction.
```

File: src/Discord.Net.Interactions/Attributes/CommandContextTypeAttribute.cs (hunk 8 above)
Comment (Misha-133) [inline src/Discord.Net.Interactions/Attributes/CommandContextTypeAttribute.cs:17]:
```suggestion
    /// <summary>
    ///     Gets context types this command can be executed in.
    /// </summary>
    public IReadOnlyCollection<InteractionContextType> ContextTypes { get; }
```

File: src/Discord.Net.Interactions/Attributes/CommandContextTypeAttribute.cs (hunk 9 above)
Comment (Misha-133) [inline src/Discord.Net.Interactions/Attributes/CommandContextTypeAttribute.cs:23]:
```suggestion
    /// <summary>
    ///     Sets the <see cref="IApplicationCommandInfo.ContextTypes"/> property of an application command or module.
    /// </summary>
    /// <param name="contextTypes">Context types set for the command.</param>
    public CommandContextTypeAttribute(params InteractionContextType[] contextTypes)
```

File: src/Discord.Net.Interactions/Attributes/IntegrationTypeAttribute.cs (hunk 10 above)
Comment (Misha-133) [inline src/Discord.Net.Interactions/Attributes/IntegrationTypeAttribute.cs:17]:
```suggestion
    /// <summary>
    ///     Gets integration install types for this command.
    /// </summary>
    public IReadOnlyCollection<ApplicationIntegrationType> IntegrationTypes { get; }
```

File: src/Discord.Net.Interactions/Attributes/IntegrationTypeAttribute.cs (hunk 11 above)
Comment (Misha-133) [inline src/Discord.Net.Interactions/Attributes/IntegrationTypeAttribute.cs:23]:
```suggestion
    /// <summary>
    ///     Sets the <see cref="IApplicationCommandInfo.IntegrationTypes"/> property of an application command or module.
    /// </summary>
    /// <param name="integrationTypes">Integration install types set for the command.</param>
    public IntegrationTypeAttribute(params ApplicationIntegrationType[] integrationTypes)
```

**Pull request merged**